homeassistant/components/vallox/sensor.py

```diff
@@ -264,16 +117,131 @@ class ValloxFilterRemainingSensor(ValloxSensor):
     async def async_update(self):
         """Fetch state from the ventilation unit."""
         try:
-            days_remaining = int(self._state_proxy.fetch_metric(self._metric_key))
+            days_remaining = int(
+                self._state_proxy.fetch_metric(self.entity_description.metric_key)
+            )
             days_remaining_delta = timedelta(days=days_remaining)
 
             # Since only a delta of days is received from the device, fix the
             # time so the timestamp does not change with every update.
             now = datetime.utcnow().replace(hour=13, minute=0, second=0, microsecond=0)
 
-            self._state = (now + days_remaining_delta).isoformat()
-            self._available = True
+            self._attr_native_value = (now + days_remaining_delta).isoformat()
+            self._attr_available = True
 
         except (OSError, KeyError) as err:
-            self._available = False
+            self._attr_available = False
             _LOGGER.error("Error updating sensor: %s", err)
+
+
+@dataclass
+class ValloxOptionalKeysMixin:
+    """Mixin for required keys."""
+
+    metric_key: str | None = None
+    sensor_type: type[ValloxSensor] = ValloxSensor
+
+
+@dataclass
+class ValloxSensorEntityDescription(ValloxOptionalKeysMixin, SensorEntityDescription):
+    """Describes Vallox sensor entity."""
+
+
+SENSORS: tuple[ValloxSensorEntityDescription, ...] = (
```

Comment:
I would recommend to move these back to the top like it's done in nearly all other integrations (after `_LOGGER`). Similar for `async_setup_platform` below.

Reply:
I got complaints from the type checker that `ValloxSensor` and its child classes aren't defined yet. This was the reason I moved it.

What's the correct way to solve this?

Reply:
It's a bit more work, but the best way would probably be to create individual description tuples for each sensor class and not store `sensor_type` in the description itself.

Reply:
With that it will also be a bit more obvious which description results in which sensor. Currently it's a bit more searching.
Another advantage of that approach would be that you could customize the `init` arguments for each sensor class if they should change in the future.

Reply:
Hmm, I must say I find the current approach quite elegant. Individual description tuples would mean four tuples of which three only contain a single entry.

How about adding a `sensor_impl.py` and importing from there?

Reply:
> How about adding a `sensor_impl.py` and importing from there?

Doesn't that create a circular dependency?

> Hmm, I must say I find the current approach quite elegant. Individual description tuples would mean four tuples of which three only contain a single entry.

I do agree here. Although I would still think individual once would be more readable.
AFAIK there is no formal rule that `async_setup_platform` must be placed at the top. I you prefer the way it's now, I won't stand in your way 🙂

Reply:
> Wouldn't that create a circular dependency?

You are right, it doesn't work.

> I you prefer the way it's now, I won't stand in your way 🙂

In this case, I would like to leave it as is, unless I face more opposition to it. 

Thanks for all your feedback!